wallet: fix unrelated parent conflict doesn't cause child tx to be marked as conflict

Eunovo opened the pull request:
This PR implements a fix for the issue described in https://github.com/bitcoin/bitcoin/issues/29435.

The problem is that the wallet is unable to abandon transactions that have unrelated parent conflicts. The solution implemented here, augments the mempool transaction `REPLACED` signal with the double-spending transaction which the wallet stores and watches for in Block notifications. A map is added to the wallet to track conflicting tx ids and their child transactions.  The entry is erased when the double-spending tx is removed from MemPool.

achow101 committed:
test: Conflict with unconfirmed parent not in wallet

Eunovo committed:
Change MemPoolRemovalReason to variant type

This allows the mempool to send additional data with TransactionRemovedFromMempool event.
Now, we can send conflicting_block_hash and conflicting_block_height for Conflicts and replacement_tx for Replacements.

Eunovo committed:
Handle double-spending of unrelated parents to wallet txs

Detect replacement of wallet txs and wait for confirmation of replacement tx before marking wallet tx as conflicted

glozow commented:
cc @achow101?

achow101 commented:
I'm actually wondering now if it would be sufficient to just have the `MemPoolRejectReason` include the replacement txid for replacements, and the conflicting block hash for conflicts. If our tx is in the mempool, we should get these notifications when it is removed which I think would be enough without having to be looking for the parents or looking to see if we have any descendants?

Eunovo commented:
> I'm actually wondering now if it would be sufficient to just have the `MemPoolRejectReason` include the replacement txid for replacements, and the conflicting block hash for conflicts. If our tx is in the mempool, we should get these notifications when it is removed which I think would be enough without having to be looking for the parents or looking to see if we have any descendants?

@achow101 I think this makes sense. Once we add conflicting block hash for conflicts then we can safely mark wallet tx as conflicted which should solve the issue. What would we still need the replacement txid for?

achow101 commented:
> When you say "replacement txid", are you referring to the txid of the tx causing the wallet tx to kicked out?

Yes, we need the replacement txid for instances where the tx is removed by replacement rather than a block conflict.

> if so, I believe the current fix implemented here does exactly what you're describing, See https://github.com/bitcoin/bitcoin/commit/635e2e47908cbc443604e388b5a823958e5b5e30

It appears to also be watching for replacements of those replacements too, and I think that is unnecessary.

Eunovo commented:
> It appears to also be watching for replacements of those replacements too, and I think that is unnecessary.

Thanks @achow101. Yes, I added that when I realized that if the replacement is also replaced then the check in the `BlockConnected` callback will fail to mark the wallet tx as conflicted. Unless for some reason, the replacement cannot be replaced?

Eunovo committed:
Handle new blocks with unrelated parents conflicts

Watch for wallet transaction conflicts triggered by adding conflicting blocks

Eunovo commented:
Rebased https://github.com/bitcoin/bitcoin/commit/532d25fb536da953db2b7a6ce7405a4b105c1a56 to https://github.com/bitcoin/bitcoin/commit/47750d3cea9baf9538925362608f9a039bc760ff 

> It appears to also be watching for replacements of those replacements too, and I think that is unnecessary.

@achow101 I took this out because the new replacement is not guaranteed to conflict with the original wallet transaction

Added `conflicting_block_hash` and `conflicting_block_height` to `ConflictReason` in https://github.com/bitcoin/bitcoin/pull/29680/commits/8b5d3d72fa54b9e35507739f53d158f5f46f05ac and 
used this information to mark wallet tx has conflicted in https://github.com/bitcoin/bitcoin/pull/29680/commits/47750d3cea9baf9538925362608f9a039bc760ff.

I had to use `RecursiveUpdateTxState` directly in https://github.com/bitcoin/bitcoin/pull/29680/commits/47750d3cea9baf9538925362608f9a039bc760ff because `CWallet::MarkConflicted` checks that the conflicting block height is not more than the `m_last_block_processed` by the wallet but `transactionRemovedFromMempool` is triggered before `blockConnected` so the wallet hasn't had a chance to process the block causing the conflict notification. I had to force the wallet txs update. _I wonder what the repercussions for doing this are_.

Curious to see what others think.

Eunovo commented:
Putting in draft while I fix falling test

josibake commented:
Concept ACK

Seems like a reasonable approach to the problem described in #29435 , will dig in more. Using a variant as a replacement for the Enum seems a bit odd at first glance and requires a lot of duplicate code for each struct.  Perhaps another approach could be a class for `RemovalReason`, which contains the Enum as a field and then uses a std::variant to hold the data needed by the wallet, if any. Would be more extensible, and then the to string method could be defined once on the class. Special logic per reason can also be handled all in the same place by switching on the Enum value and taking appropriate action per reason.

Eunovo commented:
>  Perhaps another approach could be a class for `RemovalReason`, which contains the Enum as a field and then uses a std::variant to hold the data needed by the wallet, if any. Would be more extensible, and then the to string method could be defined once on the class. Special logic per reason can also be handled all in the same place by switching on the Enum value and taking appropriate action per reason.

Thank you @josibake. Will this approach still require the definition of structs to hold each data for each reason?

Eunovo commented:
> I don't think so: only the Conflict and Replace reasons need extra data, and from what I understand it can only be one reason at any given time. Given that, you could have a field on the new class for removal data which is a std::variant of the possible data types. For example, `std::variant<CTransactionRef, BlockData>`, where block data holds the block hash and block number.

@josibake Makes sense but I'm skeptical about how this affects the removal reason usage. For example, you can just define `SizeLimitReason()` right now, but with this approach, you'd have to do  `RemovalReason(RemovalReasons::SIZE_LIMIT)` and this new approach still requires the use of `std::variant`, so, I'm worried that the current code in https://github.com/bitcoin/bitcoin/pull/29680/commits/59ee08e43255bf0e96d1fab8eb243a4b8b7864c2 will still be more clear.
WDYT?

josibake commented:
The benefits I see are we can keep the existing Enum and code for converting the Enum to a string. This would also allow updating the function signatures with a scripted-diff, which makes larger refactors like this easier to verify as a reviewer.

It also feels like a more natural fit to me and my gut feeling is that having a class to encapsulate the removal reasons and necessary data will be more maintainable and extensible going forward.

Eunovo commented:
> The benefits I see are we can keep the existing Enum and code for converting the Enum to a string. This would also allow updating the function signatures with a scripted-diff, which makes larger refactors like this easier to verify as a reviewer.
> 
> It also feels like a more natural fit to me and my gut feeling is that having a class to encapsulate the removal reasons and necessary data will be more maintainable and extensible going forward.

@josibake Makes sense. I'll create a RemovalReason class

ryanofsky commented:
> Given that, you could have a field on the new class for removal data which is a std::variant of the possible data types. For example, `std::variant<CTransactionRef, BlockData>`, where block data holds the block hash and block number.

The drawback of a separate data field is that there is no longer a guarantee that data accompanying the reason is always present when the reason is set, and always absent when it is not set. So it is possible for code to only partially initialize the RemovalReason class or initialize it in an inconsistent state.

In general I like the idea of replacing enums with variants for more safety, and to be able to express rules about state in type definitions. But variants in c++ are more awkward than sum types in other languages, and I did not look into this specific case, so maybe the tradeoffs in this case are not worth it.

josibake commented:
> The drawback of a separate data field is that there is no longer a guarantee that data accompanying the reason is always present when the reason is set, and always absent when it is not set. So it is possible for code to only partially initialize the RemovalReason class or initialize it in an inconsistent state.

Seems easily addressed with a constructor, no? Something like:

```c++
class RemovedReason {
public:
    MemPoolRemovalReason m_reason;
    std::variant<std::monostate, CTxReference, BlockData> m_extra_data;

    // Constructor for reasons that don't require extra data
    RemovedReason(MemPoolRemovalReason r) : reason(r) {
        if (requiresExtraData(r)) {
            throw std::invalid_argument("reason X requires data Y etc");
        }
    }

    // Constructor needing CTxRef
    RemovedReason(RemovalReason r, const CTxRef& data) : reason(r), extra_data(data) {
        if (!IsA(r)) {
            throw std::invalid_argument("CTxRef is required for reason A, got bla");
        }
    }

    // Constructor needing BlockData
    RemovedReason(RemovalReason r, const BlockData& data) : reason(r), extra_data(data) {
        if (!IsB(r)) {
            throw std::invalid_argument("BlockData is required for reason B, got bla");
        }
    }
```

Maybe this is starting to be more complicated than just having a struct per each reason? But I'd still argue this is a better approach in that it keeps all the logic for mempool removal reasons in one place and avoids duplicating code on each struct.

Eunovo commented:
> Maybe this is starting to be more complicated than just having a struct per each reason? But I'd still argue this is a better approach in that it keeps all the logic for mempool removal reasons in one place and avoids duplicating code on each struct.

Same thing I was thinking. Using the class right now looks like it will make things more complicated. Maybe we should leave the class for a future change where it becomes necessary? @josibake @ryanofsky

ryanofsky commented:
> Seems easily addressed with a constructor, no? Something like:

Yes, but those are manual constraints that you are writing by hand rather than automatic constraints expressed implicitly in the data definition. Depending on the constructors it may also only provide runtime checking rather than compile-time checking like in your example. And if the struct is mutable could allow invalid representations of state after construction.

I don't know what is best in this particular case, I would just stand up for:

```c++
struct MyState1 { int data; };
struct MyState2 { bool flag; };
using MyState = std::variant<MyState1, MyState2>;
```

as a good alternative to:

```c++
enum class MyState {
  STATE1,
  STATE2,
};

class MyData {
  MyState m_state,
  // ... more data and methods...
};
```

in many cases.

> Maybe we should leave the class for a future change where it becomes necessary?

I'm not sure the answer to this, but it is probably worth experimenting and choosing the approach that seems simplest.

josibake commented on src/kernel/mempool_removal_reason.h:
```diff
@@ -5,20 +5,61 @@
 };
 
+struct ConflictReason {
+    uint256 conflicting_block_hash;
+    unsigned int conflicting_block_height;
```
in https://github.com/bitcoin/bitcoin/pull/29680/commits/7debac0b09faf051229a4ead0b57e6658e00ac11 ("Change MemPoolRemovalReason to variant type"):

I haven't finished reviewing the later commits yet, but seems odd to pass both `conflicting_block_hash` and `conflicting_block_height`. Seems like we should be able to only use `conflicting_block_hash`?

Eunovo replied:
The `TxStateBlockConflicted` requires both that's why I added the both of them

josibake reviewed (commented):
Still reviewing the later commits, but had some initial feedback/questions for the first commit.

Eunovo commented on src/validation.cpp:
```diff
@@ -3057,7 +3057,7 @@ bool Chainstate::ConnectTip(BlockValidationState& state, CBlockIndex* pindexNew,
     // Remove conflicting transactions from the mempool.;
     if (m_mempool) {
-        m_mempool->removeForBlock(blockConnecting.vtx, pindexNew->nHeight);
+        m_mempool->removeForBlock(blockConnecting.vtx, pthisBlock->GetHash(), pindexNew->nHeight);
```
@josibake I looked around a bit and there might be something here.  `GetHash()` is called before this point
- when we read the block from disk https://github.com/bitcoin/bitcoin/blob/e163d864d380956a4c0f89a4d80a76f5aefc9a08/src/node/blockstorage.cpp#L1092
- when `ActivateBestChain` is called https://github.com/bitcoin/bitcoin/blob/e163d864d380956a4c0f89a4d80a76f5aefc9a08/src/validation.cpp#L3354

There might be some gain in caching GetHash() but I think that has to be addressed on its own. I'll have to measure the runtimes and see its worth a PR.

I did also discover that `pindexNew` which is a `CBlockIndex` does store the `blockhash`, see https://github.com/bitcoin/bitcoin/blob/e163d864d380956a4c0f89a4d80a76f5aefc9a08/src/chain.h#L244 Looks like I can use that here to prevent recalculating the blockhash.

furszy replied:
Can use `pindexNew` hash. We cache the block hash there.

Eunovo commented on src/kernel/mempool_removal_reason.h:
```diff
@@ -5,20 +5,61 @@
+    uint256 conflicting_block_hash;
+    unsigned int conflicting_block_height;
+
+    explicit ConflictReason(const uint256& conflicting_block_hash, int conflicting_block_height) : conflicting_block_hash(conflicting_block_hash), conflicting_block_height(conflicting_block_height) {}
+    std::string toString() const noexcept {
+        return "conflict";
+    }
+};
+
+struct ReplacedReason {
+    CTransactionRef replacement_tx;
+
+    explicit ReplacedReason(const CTransactionRef replacement_tx) : replacement_tx(replacement_tx) {}
```
> Any reason why `conflicting_block_hash` is passed by reference but `replacement_tx` isn't?

`CTransactionRef` is already a pointer https://github.com/bitcoin/bitcoin/blob/e163d864d380956a4c0f89a4d80a76f5aefc9a08/src/primitives/transaction.h#L423-L424

Eunovo replied:
> Would also be nice if we prevented these objects from be constructed with `nullptrs`. I'm not sure if we have a convention in our codebase around this or other examples to point to, but would be worth looking into.

It looks like I may be able to do this by deleting the constructor
```
// Deleted constructor
ReplacedReason(std::nullptr_t) = delete;
```
I'll test it

Eunovo replied:
@josibake Deleting the nullptr constructor worked.

furszy commented on src/wallet/wallet.cpp:
```diff
@@ -1440,15 +1440,15 @@ void CWallet::transactionAddedToMempool(const CTransactionRef& tx) {
     // Handle transactions that were removed from the mempool because they
     // conflict with transactions in a newly connected block.
-    if (reason == MemPoolRemovalReason::CONFLICT) {
+    if (std::get_if<ConflictReason>(&reason)) {
```
Could replace all the `std::get_if` for a more readable `IsReason<Conflict>(&reason)` if you include this function:
```c++
template <typename T>
bool IsReason(const MemPoolRemovalReason& reason)
{
    return std::get_if<T>(reason) != nullptr;
}
```

Eunovo replied:
Thanks @furszy

furszy commented on src/validation.cpp:
```diff
@@ -1236,7 +1236,7 @@ bool MemPoolAccept::Finalize(const ATMPArgs& args, Workspace& ws)
         m_subpackage.m_replaced_transactions.push_back(it->GetSharedTx());
     }
-    m_pool.RemoveStaged(m_subpackage.m_all_conflicts, false, MemPoolRemovalReason::REPLACED);
+    m_pool.RemoveStaged(m_subpackage.m_all_conflicting, false, ReplacedReason(entry->GetSharedTx()));
```
In b67c82a6f2ab6:

This commit does not compile, shouldn't have changed `m_all_conflicts` for `m_all_conflicting`.

Eunovo commented on src/wallet/wallet.h:
```diff
@@ -56,7 +56,6 @@ class CKeyID;
 class CPubKey;
 class Coin;
 class SigningProvider;
-enum class MemPoolRemovalReason;
```
I followed IWYU here and added kernel/mempool_removal_reason.h to wallet.h

Eunovo commented:
Thanks for the reviews @josibake  and @furszy I have implemented your suggested changes

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -1452,6 +1464,24 @@ void CWallet::transactionRemovedFromMempool(const CTransactionRef& tx, const Mem
+    auto* replaced_reason = std::get_if<ReplacedReason>(&reason);
+
+    // Check if wallet transaction is being replaced by an unrelated parent transaction
+    if (replaced_reason != nullptr && IsFromMe(*tx) && !IsFromMe(*(replaced_reason->replacement_tx))) {
```
In 7eba56accbd1b4ce3c09656c8d69b5aa437a706f "Handle double-spending of unrelated parents to wallet txs"

The replacement could still be sending to us, and in that case, it isn't unrelated.

Eunovo replied:
Updated comment in https://github.com/bitcoin/bitcoin/pull/29680/commits/3c893c51426b2bd9b9b504d00614f3f1742322d5

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -1452,6 +1464,24 @@ void CWallet::transactionRemovedFromMempool(const CTransactionRef& tx, const Mem
+        m_unrelated_conflict_tx_watchlist.insert(std::make_pair(replaced_reason->replacement_tx->GetHash(), tx));
+    }
+
+    {
```
In 7eba56accbd1b4ce3c09656c8d69b5aa437a706f "Handle double-spending of unrelated parents to wallet txs"

The extra scoping is not necessary, just name `it` something else.

Eunovo replied:
Fixed in https://github.com/bitcoin/bitcoin/pull/29680/commits/3c893c51426b2bd9b9b504d00614f3f1742322d5

achow101 commented on src/wallet/wallet.h:
```diff
@@ -428,6 +428,11 @@ class CWallet final : public WalletStorage, public interfaces::Chain::Notificati
     std::unordered_map<CScript, std::vector<ScriptPubKeyMan*>, SaltedSipHasher> m_cached_spks;
 
+    /**
+     * Tracks txids of unrelated double-spending txs  to wallet transactions
```
In 7eba56accbd1b4ce3c09656c8d69b5aa437a706f "Handle double-spending of unrelated parents to wallet txs"

nit: whitespace.

Also, I think it would be better to clarify that the key is the txid of the replacing transaction which may be unrelated to the wallet, and the value is the replaced transaction which is related to this wallet.

Eunovo replied:
Fixed in https://github.com/bitcoin/bitcoin/pull/29680/commits/3c893c51426b2bd9b9b504d00614f3f1742322d5

Eunovo commented:
Rebased https://github.com/bitcoin/bitcoin/commit/19f39271fa9f5b26075ca8f66afe25751813bb75 to https://github.com/bitcoin/bitcoin/commit/ef7b1e8e82819e5104c1f1f49f46ea74b416ced0

>This PR implements a fix for the issue described in https://github.com/bitcoin/bitcoin/issues/29435.
>
>The problem is that the wallet is unable to abandon transactions that have unrelated parent conflicts. The solution implemented here augments the mempool transaction REPLACED signal with the double-spending transaction which the wallet stores and watches for in Block notifications. A map is added to the wallet to track conflicting tx ids and their child transactions. The entry is erased when the double-spending tx is removed from MemPool.

https://github.com/bitcoin/bitcoin/pull/30072 introduced Mempool Changesets for packages and single transactions. Previously, I captured the replacement_tx when a tx was being kicked out of the Mempool, but with the addition of packages, any of the tx in the package could be the reason for the conflict.
I've changed the replacement_tx to be the last child tx in the package so the wallet doesn't have to watch the entire replacement package.

This solution assumes that all transactions in a package will be added to a Block. Suppose the first transaction in a 1P1C package is the actual replacement tx and the package only partially gets added to a Block. In that case, the wallet will not abandon the conflicted wallet tx because it's watching the child of the package.

Eunovo commented:
Putting in draft while I fix failing CI

ismaelsadeeq commented on src/interfaces/chain.h:
```diff
@@ -7,6 +7,7 @@
 
```
In 4c64dd5a3cc4a48e5470fff26c99dde20f81c7e0   "Change MemPoolRemovalReason to variant type"
nit:  Commit message should follow the guidelines in  https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#committing-patches

The explanation is a long should break it.

ismaelsadeeq commented on src/kernel/mempool_removal_reason.h:
```diff
@@ -5,20 +5,77 @@
+    uint256 conflicting_block_hash;
+    unsigned int conflicting_block_height;
+
+    explicit ConflictReason(const uint256& conflicting_block_hash, int conflicting_block_height) : conflicting_block_hash(conflicting_block_hash), conflicting_block_height(conflicting_block_height) {}
```
In 4c64dd5a3cc4a48e5470fff26c99dde20f81c7e0   "Change MemPoolRemovalReason to variant type"
nit:  this line is too long, please break.

ismaelsadeeq commented on src/wallet/wallet.cpp:
```diff
@@ -1464,6 +1476,22 @@ void CWallet::transactionRemovedFromMempool(const CTransactionRef& tx, const Mem
         RefreshMempoolStatus(it->second, chain());
     }
+
+    auto* replaced_reason = std::get_if<ReplacedReason>(&reason);
+
+    // Check if wallet transaction is being replaced by a parent transaction which is not from this wallet
+    if (IsFromMe(*tx) && replaced_reason != nullptr && replaced_reason->replacement_tx.has_value() && !IsFromMe(*replaced_reason->replacement_tx.value())) {
+        m_unrelated_conflict_tx_watchlist.insert(std::make_pair(replaced_reason->replacement_tx.value()->GetHash(), tx));
+    }
+
+    auto iter = m_unrelated_conflict_tx_watchlist.find(tx->GetHash());
+    if (iter != m_unrelated_conflict_tx_watchlist.end()) {
+        // The replacement tx was removed from the mempool, remove it from map
+        // This new replacement tx may not conflict with the original tx
+        // so leave wallet tx to remain as TxStateInactive
+        m_unrelated_conflict_tx_watchlist.erase(iter);
+    }
```
In 71200b72ab848c3066e5123dc0badc24c9648f47 "Handle double-spending of unrelated parents to wallet txs"
```suggestion
    auto* replaced_reason = std::get_if<ReplacedReason>(&reason);
    if (IsFromMe(*tx) && replaced_reason != nullptr) {

        // Check if wallet transaction is being replaced by a parent transaction which is not from this wallet
        if (replaced_reason->replacement_tx.has_value() && !IsFromMe(*replaced_reason->replacement_tx.value())) {
            m_unrelated_conflict_tx_watchlist.insert(std::make_pair(replaced_reason->replacement_tx.value()->GetHash(), tx));
        }
        // Remove the replacement tx was removed from the mempool, remove it from map
        // This new replacement tx may not conflict with the original tx
        // so leave wallet tx to remain as TxStateInactive
        m_unrelated_conflict_tx_watchlist.erase(tx->GetHash());
    }
```

ismaelsadeeq commented on src/wallet/wallet.cpp:
```diff
@@ -1464,6 +1476,22 @@ void CWallet::transactionRemovedFromMempool(const CTransactionRef& tx, const Mem
+    auto iter = m_unrelated_conflict_tx_watchlist.find(tx->GetHash());
+    if (iter != m_unrelated_conflict_tx_watchlist.end()) {
+        // The replacement tx was removed from the mempool, remove it from map
+        // This new replacement tx may not conflict with the original tx
+        // so leave wallet tx to remain as TxStateInactive
```
In 71200b72ab848c3066e5123dc0badc24c9648f47 "Handle double-spending of unrelated parents to wallet txs"

How do you know that without checking ?

ismaelsadeeq reviewed (commented):
Concept ACK

Left some initial comments

> Previously, I captured the replacement_tx when a tx was being kicked out of the Mempool, but with the addition of packages, any of the tx in the package could be the reason for the conflict.
I've changed the replacement_tx to be the last child tx in the package so the wallet doesn't have to watch the entire replacement package.

Can you elaborate on how that work, and the wallet detects that, what if the conflict is not from the child, it's from a parent transaction?

> This solution assumes that all transactions in a package will be added to a Block. 

This is Incorrect I think, you can have a package and not all the transaction will be added in the block.

> Suppose the first transaction in a 1P1C package is the actual replacement tx and the package only partially gets added to a Block. 
In that case, the wallet will not abandon the conflicted wallet tx because it's watching the child of the package.

How is this not an issue?

Eunovo commented:
@ismaelsadeeq I wanted to get other opinions on this. With the addition of changesets, the problem is more complicated. 

> Can you elaborate on how that work, and the wallet detects that, what if the conflict is not from the child, it's from a parent transaction?

It does work because it's guaranteed that the parent will be confirmed before the child. The wallet can mark the replaced wallet tx as conflicted when the child of the package gets confirmed. The problem though is that the child may never be confirmed even if the parent has been confirmed.

> This is Incorrect I think, you can have a package and not all the transaction will be added in the block.
> How is this not an issue?

It is.
